Move "Whitespace" article from DOM to CSS (#40523)
mdn/content: files/en-us/web/css/css_text/whitespace/index.md

Note that this list does not include non-breaking spaces (U+00A0, `&nbsp;` in HTML). So these characters don't trigger any [collapsing](/en-US/docs/Web/CSS/CSS_text/Whitespace#collapsing_and_transformation), which is why they are often used to create longer spaces in HTML.

## How does HTML process whitespace?

**HTML generally does not ignore whitespace.** As a markup language, HTML produces a {{glossary("DOM")}} where all whitespace in the text content is preserved. This is how CSS, a downstream rendering engine that works on the DOM, lets us control how whitespace is displayed using the {{cssxref("white-space")}} property. Every single character of text content in the source code is represented in the DOM tree and can be retrieved and manipulated via DOM APIs such as {{domxref("Node.textContent")}}.

> To be clear, we're talking about whitespace _between HTML tags_, which becomes text nodes in the DOM. Any whitespace _inside a tag_ (between the angle brackets but not as part of an attribute value) is just part of the HTML syntax and does not appear in the DOM.

Author: ```suggestion
> [!NOTE]
> To be clear, we're talking about whitespace _between HTML tags_, which becomes text nodes in the DOM. Any whitespace _inside a tag_ (between the angle brackets but not as part of an attribute value) is just part of the HTML syntax and does not appear in the DOM.
```

Different {{cssxref("white-space-collapse")}} values skip different steps of this algorithm:

- `preserve` and `break-spaces`: the whole algorithm except step 3 is skipped, and no whitespace collapsing or transformation happens.
- `preserve-spaces`: the whole algorithm is skipped, so whitespace at the start and end of lines are preserved.
- `preserve-breaks`: the algorithm is the same as `collapse`.

### Whitespace processing between inline and inline-block elements

Let's move on to look at a few issues that can arise due to whitespace, and what can be done about them. First of all, we'll look at what happens with spaces in between inline and inline-block elements. In fact, we saw this already in our very first example, when we described how whitespace is processed inside inline formatting contexts.

> We said that there were rules to ignore most characters but that word-separating characters remain. When you're only dealing with block-level elements such as `<p>` that only contain inline elements such as `<em>`, `<strong>`, `<span>`, etc., you don't normally care about this because the extra whitespace that does make it to the layout is helpful to separate the words in the sentence.
> 
> It gets more interesting however when you start using `inline-block` elements. These elements behave like inline elements on the outside, and blocks on the inside, and are often used to display more complex pieces of UI than just text, side-by-side on the same line, for example navigation menu items.
> 
> Because they are blocks, many people expect that they will behave as such, but really they don't. If there is formatting whitespace between adjacent inline elements, this will result in space in the layout, just like the spaces between words in text.

Reviewer: Can we organize all of this content into points, specifically some that has already been discussed and remaining that is new. Also remove any extraneous words to keep the content compact and crisp since the page is already very long
```suggestion
As we saw earlier with inline and block elements:
- Most whitespace characters are ignored, but word-separating characters like spaces remain.
- The extra whitespace that does make it to the layout is helpful to separate the words in the sentence.

With `inline-block` elements, it gets more interesting:
- These elements behave like inline elements on the outside and blocks on the inside. (They're often used to display more complex pieces of UI, placed side by side on the same line, such as navigation menu items.)
- Any whitespace between adjacent inline or inline-block elements will result in spaces in the layout, just like the spaces between words in text. (This can surprise developers because they are blocks, and blocks don't normally show extra spaces.) 
```

Consider this example (again, we've included an HTML comment that shows the whitespace characters in the HTML):

```css live-sample___inline-block
.people-list {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.people-list li {
  display: inline-block;
  width: 2em;
  height: 2em;
  background: #f06;
  border: 1px solid;
}
```

```html live-sample___inline-block
<ul class="people-list">
  <li></li>

  <li></li>

  <li></li>

  <li></li>

  <li></li>
</ul>

<!--
<ul class="people-list">⏎
◦◦<li></li>⏎
⏎
◦◦<li></li>⏎
⏎
◦◦<li></li>⏎
⏎
◦◦<li></li>⏎
⏎
◦◦<li></li>⏎
</ul>
-->
```

This renders as follows:

{{EmbedLiveSample('inline-block')}}

> You probably don't want the gaps in between the blocks — depending on the use case (is this a list of avatars, or horizontal nav buttons?), you probably want the element sides flush with each other, and to be able to control any spacing yourself.

Reviewer: ```suggestion
You probably don't want the gaps between the blocks. Depending on your use case (such as a list of avatars or a horizontal row of navigation buttons), you probably want the elements to sit flush against each other, and to be able to control any spacing yourself.
```

> The Firefox DevTools HTML Inspector will highlight text nodes, and also show you exactly what area the elements are taking up — useful if you are wondering what is causing the problem, and are maybe thinking you've got some extra margin in there or something!

Reviewer: ```suggestion
The Firefox DevTools HTML Inspector can highlight text nodes and also show you exactly the area the elements are taking up. This is useful to check out if you suspect there's extra margin or unexpected whitespace causing gaps.
```

![Example of displaying whitespaces between blocks in the Firefox DevTools HTML Inspector](whitespace-devtools.png)

There are a few ways of getting around this problem:

> Use [Flexbox](/en-US/docs/Learn_web_development/Core/CSS_layout/Flexbox) to create the horizontal list of items instead of trying an `inline-block` solution. This handles everything for you, and is definitely the preferred solution:

Reviewer: since we're talking about the multiple ways, let's make them bulleted
```suggestion
- Use [Flexbox](/en-US/docs/Learn_web_development/Core/CSS_layout/Flexbox) to create the horizontal list of items instead of trying an `inline-block` solution. Flexbox handles spacing and alignment for you, and is definitely the preferred solution:
```

```css
ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
  display: flex;
}
```

> If you need to rely on `inline-block`, you could set the [`font-size`](/en-US/docs/Web/CSS/font-size) of the list to 0. This only works if your blocks are not sized with ems (based on the `font-size`, so the block size would also end up being 0). rems would be a good choice here:

Reviewer: ```suggestion
- If you need to rely on `inline-block`, you could set the [`font-size`](/en-US/docs/Web/CSS/font-size) of the list to `0`. This only works if the blocks are not sized with `em` units (since `em` is based on the `font-size`, so the block size would also end up being sized as `0`). Using `rem` units would be a good choice here:
```

```css
ul {
  font-size: 0;
  /* … */
}

li {
  display: inline-block;
  width: 2rem;
  height: 2rem;
  /* … */
}
```

> Or you could set negative margin on the list items:

Reviewer: ```suggestion
Alternatively, you could set negative margin on the list items:
```

```css
li {
  display: inline-block;
  width: 2rem;
  height: 2rem;
  margin-right: -0.25rem;
}
```

> You can also solve this problem by avoiding whitespace nodes between `<li>` items:

Reviewer: ```suggestion
- You can also solve this problem by avoiding whitespace nodes between `<li>` items:
```

```html-nolint
<li>
  ...
</li><li>
  ...
</li>
```

## Working with whitespace in the DOM

> As mentioned previously, whitespace is collapsed and trimmed when rendered, but preserved in DOM. This may present some pitfalls when trying to do [DOM](/en-US/docs/Web/API/Document_Object_Model) manipulation in JavaScript. For example, if you have a reference to a parent node and want to affect its first element child using [`Node.firstChild`](/en-US/docs/Web/API/Node/firstChild), if there is a rogue whitespace node just after the opening parent tag you will not get the result you are expecting. The text node would be selected instead of the element you want to affect.

Reviewer: ```suggestion
As mentioned previously, whitespace is [collapsed and trimmed](#collapsing_and_transformation) when rendered, but preserved in the DOM. This may present some pitfalls when trying to do [DOM](/en-US/docs/Web/API/Document_Object_Model) manipulation in JavaScript. For example, if you have a reference to a parent node and want to affect its first element child using [`Node.firstChild`](/en-US/docs/Web/API/Node/firstChild), a rogue whitespace node just after the opening parent tag you will not give you the result you're expecting. The text node would be selected instead of the element you want to target.
```

> As another example, if you have a certain subset of elements that you want to do something to based on whether they are empty (have no child nodes) or not, you could check whether each element is empty using something like [`Node.hasChildNodes()`](/en-US/docs/Web/API/Node/hasChildNodes), but again, if any target elements contain text nodes, you could end up with false results.

Reviewer: ```suggestion
As another example, if you want to do something to a subset of elements based on whether they are empty (have no child nodes), you could use [`Node.hasChildNodes()`](/en-US/docs/Web/API/Node/hasChildNodes). But if any of those elements contain text nodes, you could end up with false results.
```

> The JavaScript code below defines several functions that make it easier to deal with whitespace in the DOM:

Reviewer: ```suggestion
The following JavaScript code block shows several functions that make it easier to deal with whitespace in the DOM:
```

   ```html-nolint
   <h1>◦◦◦Hello◦<span>◦World!</span>◦◦◦</h1>
   ```

   Notably, in languages that have no word separators, such as Chinese, lines are joined with no intervening space. So:

   ```html-nolint
   <div>你好
   世界</div>
   ```

   might be rendered as "你好世界" without any spaces in between, depending on the browser's heuristics.

4. Next, all tab characters are transformed into space characters, so the example becomes:

   ```html-nolint
   <h1>◦◦◦Hello⏎
   <span>◦World!</span>◦◦◦</h1>
   ```

5. After that, any space immediately following another space (even across two separate inline elements) is ignored, so we end up with:

   ```html-nolint
   <h1>◦Hello◦<span>World!</span>◦</h1>
   ```

This is why people visiting the web page will see the phrase "Hello World!" nicely written at the top of the page, rather than a weirdly indented "Hello" followed by an even more weirdly indented "World!" on the next line.

> After these steps, the browser processes line wrapping and bidirectional text, which we will ignore here. Note that there are still spaces left after the opening `<h1>` tag and before the closing `</h1>` tag, but these are not rendered in the browser. We'll handle that next, as each line is laid out.

Author: This is intended. These paragraphs wrap up the algorithm and aren't intended to be specific to step 5.

   </body>
   ```

3. Each tab that's preserved at this point is rendered according to {{cssxref("tab-size")}}. This can only happen with `white-space-collapse` set to `preserve` or `break-spaces` because all other settings turn tabs into spaces.
4. Sequences of spaces at the end of a line are removed, so the above becomes:

   ```html-nolint
   <body>

   <div>Hello</div>

   <div>World!</div>

   </body>
   ```

The three empty lines we now have are not going to occupy any space in the final layout, because they don't contain any visible content. So we'll end up with only two lines taking up space in the page. People viewing the web page see the words "Hello" and "World!" on two separate lines, just as you'd expect two `<div>`s to be laid out. Browsers essentially ignore all of the whitespace that was included in the HTML code.

Different {{cssxref("white-space-collapse")}} values skip different steps of this algorithm:

- `preserve` and `break-spaces`: The whole algorithm is skipped except for step 3, so no whitespace collapsing or transformation happens.
- `preserve-spaces`: The whole algorithm is skipped, so whitespace characters at the start and end of lines are preserved.

> - `preserve-breaks`: the algorithm is the same as `collapse`.

Author: The `white-space-collapse` value affects both collapsing and trimming; the trimming section also starts with the note that "we assume `white-space-collapse: collapse` below".

> - `preserve-breaks`: The algorithm is the same as that of `collapse`.

Author: ```suggestion
- `preserve-breaks`: The same algorithm is applied as with the `collapse` value.
```

---
title: Handling whitespace
slug: Web/CSS/CSS_text/Whitespace
page-type: guide
sidebar: cssref
---

The presence of whitespace in the [DOM](/en-US/docs/Web/API/Document_Object_Model) can cause layout problems and make manipulation of the content tree difficult in unexpected ways, depending on where it is located. This article explores when difficulties can occur, and looks at what can be done to mitigate resulting problems.

## What is whitespace?

> _Document white space characters_, as far as the CSS whitespace processing rules are concerned, only include spaces (U+0020), tabs (U+0009), line feeds (U+000A), and carriage returns (U+000D), where CR characters are equivalent to spaces in every regard. These characters allow you to format your code for readability. In fact, much of our source code is full of these whitespace characters, and we only tend to get rid of it in a production build step to reduce code download sizes.

Author: I don't think we should explain the basics about whitespace because everyone has a basic sense that it includes spaces and tabs. We only need to clarify whether it includes line breaks or other Unicode space characters.

```suggestion
_Document white space characters_, as far as the CSS whitespace processing rules are concerned, only include spaces (U+0020), tabs (U+0009), line feeds (LF, U+000A), and carriage returns (CR, U+000D), where CR characters are equivalent to spaces in every regard. These characters allow you to format your code for readability. Much of our source code is full of these whitespace characters, and we tend to remove them only as part of a production build step to reduce file size.
```

> _Document white space characters_, as far as the CSS whitespace processing rules are concerned, only include spaces (U+0020), tabs (U+0009), line feeds (U+000A), and carriage returns (U+000D), where CR characters are equivalent to spaces in every regard. These characters allow you to format your code for readability. In fact, much of our source code is full of these whitespace characters, and we only tend to get rid of it in a production build step to reduce code download sizes.

Author: "Document white space characters" is the exact term that CSS defines.

Happy to add a "See also" but my philosophy is that the see also has to contain things not already substantially explained in the content, so I don't have great ideas about what to add: certainly not the CSS spec because it's already linked multiple times in content.

Note that because this list does not include non-breaking spaces (U+00A0, `&nbsp;` in HTML), they don't trigger any collapsing, which is why they are often used to create longer spaces in HTML.

## How does HTML process whitespace?

> **It is not true that "HTML largely ignores whitespace".** HTML, as a markup language that produces {{glossary("DOM")}}, does not ignore any whitespace in the text content—otherwise, there would be no way for CSS, a downstream rendering engine which works on the DOM, to be able to preserve whitespace via the {{cssxref("white-space")}} property! Every single character of text content in the source code corresponds to something in the DOM tree, which can be retrieved and manipulated via DOM APIs such as {{domxref("Node.textContent")}}.

Author: It's not just the previous document; it's a common myth that "XML preserves whitespace, HTML doesn't", and we need to explicitly rebut that. I agree stating it positively is better though.

> **It is not true that "HTML largely ignores whitespace".** HTML, as a markup language that produces {{glossary("DOM")}}, does not ignore any whitespace in the text content—otherwise, there would be no way for CSS, a downstream rendering engine which works on the DOM, to be able to preserve whitespace via the {{cssxref("white-space")}} property! Every single character of text content in the source code corresponds to something in the DOM tree, which can be retrieved and manipulated via DOM APIs such as {{domxref("Node.textContent")}}.

Reviewer 2: > it's a common myth that "XML preserves whitespace, HTML doesn't"

Some people will have heard that, lots won't. If someone hasn't, it comes across very weird. So if you want to rebut it, you should say something about it being a common myth/misconception, so the second group have this context. Personally I think it is not needed, but again this is opinion.

![The DOM tree representing a simple HTML document](dom-string.png)

Note that:

- There will be some text nodes that contain only whitespace, and
- Some text nodes will have whitespace at the beginning or end.

> Conserving whitespace characters in the DOM is useful in many ways, but there are certain places where this makes certain layouts more difficult to implement, and causes problems for developers who want to iterate through nodes in the DOM. We'll look at these, and some solutions, later on.

Author: We'll need to link to two separate places (one about layout and one about DOM iteration, both of which I'd like to keep), so I'm not keen to elaborate here.

When the DOM is passed to CSS for rendering, however, the whitespace is largely stripped by default so the code formatting becomes irrelevant to the end user.

```html-nolint
<!doctype html>

  <h1>      Hello      World!     </h1>
```

This source code contains a couple of line feeds after the `doctype` and a bunch of space characters before, after, and inside the `<h1>` element, but the browser doesn't seem to care at all and just shows the words "Hello World!" as if these characters didn't exist at all:

{{EmbedLiveSample('HTML_largely_ignores_whitespace')}}

This is so that whitespace characters don't impact the layout of your page. Creating space around and inside elements is the job of CSS.

> [!NOTE]
> Due to the magic that is HTML parsing (quote from [DOM spec](https://dom.spec.whatwg.org/#introduction-to-the-dom)), there are certain whitespace characters that could be ignored by HTML; for example, whitespace between the `<html>` and `<head>` opening tags, or `</body>` and `</html>` closing tags. Here, we just deal with concrete, rendered text content.
>

> > Furthermore, the HTML parser does [_normalize_ certain whitespaces](https://html.spec.whatwg.org/multipage/parsing.html#preprocessing-the-input-stream). It would replace CR (U+000D) and CRLF sequences with single LF (U+000A). However, CR characters can still be inserted into the DOM either via [character references](/en-US/docs/Glossary/Character_reference) or JavaScript, so the CSS whitespace processing rules still need to deal with them.

Author: The phrase "due to the magic that" is the entire reason I have it here because I find it lovely :)

> > Furthermore, the HTML parser does [_normalize_ certain whitespaces](https://html.spec.whatwg.org/multipage/parsing.html#preprocessing-the-input-stream). It would replace CR (U+000D) and CRLF sequences with single LF (U+000A). However, CR characters can still be inserted into the DOM either via [character references](/en-US/docs/Glossary/Character_reference) or JavaScript, so the CSS whitespace processing rules still need to deal with them.

Reviewer: then we should include it within quotes:
"Due to the magic that is HTML parsing" (as the [DOM specification](https://dom.spec.whatwg.org/#introduction-to-the-dom) refers to the algorithm), some whitespace characters could be ignored by HTML. For example,..." followed by my other edit suggestions

> > Furthermore, the HTML parser does [_normalize_ certain whitespaces](https://html.spec.whatwg.org/multipage/parsing.html#preprocessing-the-input-stream). It would replace CR (U+000D) and CRLF sequences with single LF (U+000A). However, CR characters can still be inserted into the DOM either via [character references](/en-US/docs/Glossary/Character_reference) or JavaScript, so the CSS whitespace processing rules still need to deal with them.

Author: It's not in quotes because it's an indirect quote (i.e., as if I'm saying it myself, except it's backed by the spec). The other suggestions look good

## How does CSS process whitespace?

> Most whitespace characters are ignored, not all of them are. In the earlier example one of the spaces between "Hello" and "World!" still exists when the page is rendered in a browser. There are rules in the browser engine that decide which whitespace characters are useful and which aren't — these are specified at least in part in the CSS Text Module, and especially the [whitespace processing details](https://drafts.csswg.org/css-text-4/#white-space-processing), but we also offer an easier explanation below.

Author: I would keep the link because the spec is not inaccessible at all, but I agree with the rest of the phrasing.

### Collapsing and transformation

Let's take an example. To make it easier, we've added a comment that shows all spaces with ◦, all tabs with ⇥, and all line breaks with ⏎:

<!-- markdownlint-disable no-hard-tabs -->

```html-nolint live-sample___ex-inline
<h1>   Hello
				<span> World!</span>	  </h1>

<!--
<h1>◦◦◦Hello◦⏎
⇥⇥⇥⇥<span>◦World!</span>⇥◦◦</h1>
-->
```

<!-- markdownlint-enable no-hard-tabs -->

This example is rendered in the browser like so:

{{EmbedLiveSample('ex-inline')}}

The `<h1>` element contains only inline elements. In fact it contains:

- A text node (consisting of some spaces, the word "Hello" and some tabs).
- An inline element (the `<span>`, which contains a space, and the word "World!").
- Another text node (consisting only of tabs and spaces).

Because of this, it establishes an [inline formatting context](/en-US/docs/Web/CSS/CSS_inline_layout/Inline_formatting_context). This is one of the possible layout rendering contexts that browser engines work with.

Inside this context, whitespace character processing can be summarized as follows:

> [!NOTE]

> > Everything below doesn't apply when {{cssxref("white-space-collapse")}} is `preserve` or `preserve-spaces`.

Author: Good point about this. I'll add different H3 sections to describe what happens with different `white-space` settings (they are much simpler to describe anyway)

> 

Author: Our live sample has a bug wherein it doesn't use the assets in this PR but the assets on main.

   <span>◦World!</span>⇥◦◦</h1>
   ```

> 2. Next, consecutive line breaks are collapsed into a single line break, but we don't have any in this example.

Author: ```suggestion
2. Next, consecutive line breaks are collapsed into a single line break. We don't have any in this example.
```

> 3. Next, any remaining segment break is either transformed into a space (U+0020) or removed depending on the context before and after the break. The rules for this operation are user-agent-defined. In our example with only English, we can expect all line breaks to be transformed into spaces, so we end up with:

Author: ```suggestion
3. Next, any remaining segment break is either transformed into a space (U+0020) or removed depending on the context before and after the break. The rules for this operation are user-agent-defined. In our example with only English, we can expect all line breaks to be transformed into spaces, so we end up with:
```

> 3. Next, any remaining segment break is either transformed into a space (U+0020) or removed depending on the context before and after the break. The rules for this operation are user-agent-defined. In our example with only English, we can expect all line breaks to be transformed into spaces, so we end up with:

Author: ```suggestion
3. Next, lines in the source code are joined into single lines by removing any remaining line break characters. They are either transformed into spaces (U+0020) or simply removed, depending on the context before and after the break. The exact choice between the two is browser- and language-dependent. In our example here in English (where spaces separate words), we can expect all line breaks to be "transformed" into spaces. So we end up with:
```

   ```html-nolint
   <h1>◦◦◦Hello◦<span>◦World!</span>◦◦◦</h1>
   ```

>    Notably, in languages that have no word separators, such as Chinese, lines are joined with no intervening space. So:

Author: "Join lines" and "remove line breaks" are synonymous. I chose the former because conceptually the line break is still there (it participates in word segmentation, is visible in copy&paste, etc.), just not going into those details

>    Notably, in languages that have no word separators, such as Chinese, lines are joined with no intervening space. So:

Reviewer: Gotcha

   ```html-nolint
   <div>你好
   世界</div>
   ```

>    might be rendered as "你好世界" without any spaces in between, depending on the user agent's heuristics.

Author: ```suggestion
   might be rendered as "你好世界" without any spaces in between, depending on the browser's heuristics.
```

   > [!NOTE]
   > Steps 2 and 3 don't happen if `white-space-collapse` is set to `preserve-breaks`.

4. Next, all tab characters are handled as space characters, so the example becomes:

   ```html-nolint
   <h1>◦◦◦Hello⏎
   <span>◦World!</span>◦◦◦</h1>
   ```

5. After that, any space immediately following another space (even across two separate inline elements) is ignored, so we end up with:

   ```html-nolint
   <h1>◦Hello◦<span>World!</span>◦</h1>
   ```

This is why people visiting the web page will see the phrase "Hello World!" nicely written at the top of the page, rather than a weirdly indented "Hello" followed but an even more weirdly indented "World!" on the line below that.

> After these steps, we process line wrapping and bidirectional text, which we will ignore here. Note we still have spaces left after the opening `<h1>` tag and before the closing `</h1>` tag, but these are not rendered in the browser. We'll handle that next, as we lay out each line.

Reviewer 2: "next" I believe refers to the "Trimming and positioning" section? It would be nice if that used the same example: it's nice in this section that each step is anchored to our example, and it's a shame we lose that (and have to start again as it were) when we go to the next section.

> [!NOTE]

> > [Firefox DevTools](https://firefox-source-docs.mozilla.org/devtools-user/index.html) supports highlighting text nodes, making it easier to see exactly what nodes whitespace characters are contained within. Pure whitespace nodes are marked with a "whitespace" label.

Reviewer 2: Not sure this is still true. I don't see the whitespace nodes in the demo: https://firefox-dev.tools/devtools-examples/whitespace-only-demo/index.html.

> In a nutshell, different whitespace characters are collapsed in the following fashion:

Reviewer 2: Honestly I don't think this "in a nutshell" bit is very useful. It recaps the previous list of steps, but in a less precise and (to me) easy to follow way. Does it add anything else?

> In a nutshell, different whitespace characters are collapsed in the following fashion:

Author: @wbamberg, This is ported from the `white-space-collapse` page, and it does present things in a case-by-case way instead of a step-by-step way which may be useful to certain people.

  cur = nodeAfter(cur);
}
```

> Not all DOM APIs preserve whitespace; some APIs deal with the _rendered text_ by design. For example, {{domxref("Element.innerText")}} returns the text exactly as it's rendered, with all whitespace collapsed and trimmed. There's another tricky API: {{domxref("Selection.toString()")}} returns the text as it would be pasted, which generally means that whitespace is collapsed. However, in Firefox (which collapses whitespace between Chinese characters, as mentioned in the [collapsing and transformation](#collapsing_and_transformation) section above), the collapsed whitespace is still preserved both in the string returned by `toString()` and in the pasted text.

Author: ```suggestion
Not all DOM APIs preserve whitespace; some APIs deal with the _rendered text_ by design. For example, {{domxref("HTMLElement.innerText")}} returns the text exactly as it's rendered, with all whitespace collapsed and trimmed. {{domxref("Selection.toString()")}} returns the text as it would be pasted, which generally means that whitespace is collapsed. However, in Firefox (which collapses whitespace between Chinese characters, as mentioned in the [collapsing and transformation](#collapsing_and_transformation) section above), the collapsed whitespace is still preserved both in the string returned by `toString()` and in the pasted text.
```
